cmd/metrics.go

```diff
@@ -100,6 +101,11 @@ func (c *minioCollector) Collect(ch chan<- prometheus.Metric) {
 		float64(globalConnStats.getTotalInputBytes()),
 	)
 
+	// don't expose disk-related metrics if an ObjectLayer doesn't provide respective info, e.g. gateways
+	if reflect.DeepEqual(s, StorageInfo{}) {
```

Comment:
We shouldn't do this inside `Collect` instead simply look for s.Backend.Type

Reply:
Sorry, I don't get it.

* Should I look for `s.Backend.Type` inside `Collect` ?
* If I start dropping disk metrics when, let's say, `if s.Backend.Type == Unknown`, then I'll make the following `else` branch inaccessible (see below). Is it ok ?

```
	var totalDisks, offlineDisks int
	// Setting totalDisks to 1 and offlineDisks to 0 in FS mode
	if s.Backend.Type == FS {
                ...
	} else {
		offlineDisks = s.Backend.OfflineDisks
		totalDisks = s.Backend.OfflineDisks + s.Backend.OnlineDisks
	}
```

In other words, the idea behind `if reflect.DeepEqual(s, StorageInfo{})` was to have a check equivalent to `if s == nil`. Which would preserve existing behaviour in a case where `s.Backend.Type == Unknown` together with `s.Backend.OfflineDisks > 0` or `s.Backend.OnlineDisks > 0`.

Reply:
The problem is the way all the metrics are listed . All networking related ones should come before the StorageInfo() since they are available always. 

There is no set order so you can make those changes ..